lib/pacemaker/pcmk_status.c

```diff
@@ -287,19 +287,8 @@ pcmk__status(pcmk__output_t *out, cib_t *cib,
     }
 
 done:
-    if (stonith != NULL) {
-        if (stonith->state != stonith_disconnected) {
-            stonith->cmds->remove_notification(stonith, NULL);
```

Comment:
It's helpful to remove all notifications *before* disconnecting; it actually unregisters the notification with the fencer, rather than just deleting the client-side notification table.

Quite possibly we should do that in stonith_api_free() (or maybe only stonith_api_delete()). Offhand, I can't think of a reason not to, except maybe to avoid a timeout if the connection is known to be not working properly.

Reply:
Would checking `stonith->state != stonith_disconnected` within free/delete before unregistering be sufficient?

Reply:
And on the fence between free and delete... at this point delete is wholly redundant to free. Could be deprecated unless we have plans for it in the future.

Reply:
> Could be deprecated

Never mind, it has the NULL check :)

Reply:
Free and delete are redundant but both are public API so we can't do anything about it :-/ Also, one returns void and the other int, which isn't significant but prevents us from at least reducing it to one actual function.

My gut feeling is that free should unregister the notifications before disconnecting, but I haven't traced through all the callers to be sure that's OK.

Reply:
The null check should moved to free, there's no reason to have that as a point of difference

Reply:
`stonith_api_free()` is static and always called as a method outside of `st_client.c`, so the argument **should** never be null there.

Also, in theory if there were more than one implementation of the free method, we could still want `stonith_api_delete()` as a wrapper that does a null check. We have other free() wrappers that just null-check scattered around IIRC.

Reply:
> My gut feeling is that free should unregister the notifications before disconnecting

I would at least *expect* that to be okay, provided the client is not already disconnected